src/mgm.rs

```diff
@@ -143,47 +147,181 @@ impl MgmAlgorithmId {
-        Self(key_bytes)
+    /// Generates a random MGM key for the given algorithm.
+    pub fn generate<C: MgmKeyAlgorithm>(rng: &mut impl CryptoRngCore) -> Result<Self> {
+        match C::ALGORITHM_ID {
+            MgmAlgorithmId::ThreeDes => MgmKey3Des::generate(rng).map(MgmKeyKind::Tdes),
+        }
+        .map(Self)
     }
 
-    /// Create an MGM key from byte slice.
+    /// Generates a random MGM key using the preferred algorithm for the given Yubikey's
+    /// firmware version.
+    pub fn generate_for(yubikey: &YubiKey, rng: &mut impl CryptoRngCore) -> Result<Self> {
+        match yubikey.version() {
+            // Initial firmware versions default to 3DES.
+            Version { major: ..=4, .. }
+            | Version {
+                major: 5,
+                minor: ..=6,
+                ..
+            } => MgmKey3Des::generate(rng).map(MgmKeyKind::Tdes),
+            // Firmware 5.7.0 and above default to AES-192.
+            Version {
+                major: 5,
+                minor: 7..,
+                ..
+            }
+            | Version { major: 6.., .. } => Err(Error::NotSupported),
+        }
+        .map(Self)
+    }
+
+    /// Parses an MGM key from the given byte slice.
     ///
     /// Returns an error if the slice is the wrong size or the key is weak.
+    ///
+    /// TODO: Can we distinguish DES from AES-192? Or do we take `C` as a parameter and
+    /// require the caller to know the type of the bytes they are parsing?
     pub fn from_bytes(bytes: impl AsRef<[u8]>) -> Result<Self> {
-        bytes.as_ref().try_into()
+        MgmKey3Des::from_bytes(bytes)
+            .map(MgmKeyKind::Tdes)
+            .map(Self)
     }
 
-    /// Create an MGM key from the given byte array.
+    /// Gets the default management key for the given Yubikey's firmware version.
     ///
-    /// Returns an error if the key is weak.
-    pub fn new(key_bytes: [u8; DES_LEN_3DES]) -> Result<Self> {
-        if tdes::is_weak_key(&key_bytes) {
+    /// Returns an error if the Yubikey's default algorithm is unsupported.
+    #[cfg(feature = "untested")]
+    pub fn get_default(yubikey: &YubiKey) -> Result<Self> {
+        match yubikey.version() {
+            // Initial firmware versions default to 3DES.
+            Version { major: ..=4, .. }
+            | Version {
+                major: 5,
+                minor: ..=6,
+                ..
+            } => Ok(Self(MgmKeyKind::Tdes(
+                MgmKey3Des::new(DEFAULT_MGM_KEY.into()).expect("valid"),
+            ))),
+            // Firmware 5.7.0 and above default to AES-192.
+            Version {
+                major: 5,
+                minor: 7..,
+                ..
+            }
+            | Version { major: 6.., .. } => Err(Error::NotSupported),
+        }
+    }
+
+    /// Derives a management key (MGM) with the given algorithm from a stored salt.
+    ///
+    /// TODO: Is this supported for AES? Is the algorithm supposed to be dynamic?
```

Comment:
No idea here

Reply:
https://docs.yubico.com/yesdk/users-manual/application-piv/pin-puk-mgmt-key.html says:
> Warning
>
> PIN-derived mode is not secure. You should not use this technique. It is offered only for backwards compatibility.

So I'm going to assume that it has never been supported for AES and only provide it for TDES.